File: typescript/src/formatter.ts
import { format } from 'date-fns';
import { utcToZonedTime } from 'date-fns-tz';
import { Config } from './config';

export class WorklogFormatter {
  private readonly config: Config;
  
  constructor(config: Config) {
    this.config = config;
  }
  
  /**
   * Format a worklog entry
   * @param date Date of the entry
   * @param hours Total hours worked
   * @param descriptions List of task descriptions
   * @param hasRunningEntry Whether there's a running entry
   * @returns Formatted worklog entry
   */
  formatEntry(
    date: Date,
    hours: number,
    descriptions: string[],
    hasRunningEntry: boolean
  ): string {
    const { timezone, logger } = this.config;
    
    // Convert to timezone
    const zonedDate = utcToZonedTime(date, timezone);
    
    // Format date as YYYY-M-D
    const dateStr = `${zonedDate.getFullYear()}-${zonedDate.getMonth() + 1}-${zonedDate.getDate()}`;
    
    // Get day name (Mon, Tue, etc.)
    const dayNames = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
    const dayName = dayNames[zonedDate.getDay()];
    
    // Format hours as X.Yh or X.Yh+ if there's a running entry
    const hoursStr = `${hours}h${hasRunningEntry ? '+' : ''}`;
    
    // Join descriptions with periods
    let descriptionText = descriptions.join('. ');
    if (descriptionText && !descriptionText.endsWith('.')) {
      descriptionText += '.';
    }
    
    // Format the entry
    return `${dateStr} ${dayName} (${hoursStr}): ${descriptionText}`;
  }
  
  /**
   * Parse a worklog entry
   * @param entry Worklog entry string
   * @returns Object with parsed components or null if parsing failed
   */
  parseEntry(entry: string): {
    date: Date;
    dayName: string;
    hours: number;
    hasRunningEntry: boolean;
    descriptions: string[];
    descriptionText: string;
  } | null {
    const { timezone, logger } = this.config;
    
    // Pattern to match worklog entries
    const pattern = /(\d{4}-\d{1,2}-\d{1,2}) ([A-Za-z]{3}) \((\d+\.?\d*)h(\+?)\): (.*)/;
    const match = entry.match(pattern);
    
    if (!match) {
      logger.warning(`Failed to parse worklog entry: ${entry}`);
      return null;
    }
    
    const [, dateStr, dayName, hoursStr, runningMarker, description] = match;
    
    try {
      // Parse date
      const [year, month, day] = dateStr.split('-').map(Number);
      const date = new Date(year, month - 1, day);
      
      // Parse hours
      const hours = parseFloat(hoursStr);
      
      // Parse running status
      const hasRunningEntry = runningMarker === '+';
      
      // Parse descriptions
      const descriptions = description
        .split('.')
        .map(d => d.trim())
        .filter(d => d.length > 0);
      
      return {
        date,
        dayName,
        hours,
        hasRunningEntry,
        descriptions,
        descriptionText: description,
      };
    } catch (error) {
      logger.warning(`Error parsing worklog entry: ${error}`);
      return null;
    }
  }
  
  /**
   * Merge an existing entry with a new entry
   * @param existingEntry Existing worklog entry
   * @param newEntry New worklog entry
   * @returns Merged worklog entry
   */
  mergeEntries(existingEntry: string, newEntry: string): string {
    const { logger } = this.config;
    
    const existingParsed = this.parseEntry(existingEntry);
    const newParsed = this.parseEntry(newEntry);
    
    if (!existingParsed || !newParsed) {
      logger.warning('Failed to parse entries for merging, using new entry');
      return newEntry;
    }
    
    // Use the higher hours value
    const hours = Math.max(existingParsed.hours, newParsed.hours);
    
    // Combine running status (if either is running, result is running)
    const hasRunningEntry = existingParsed.hasRunningEntry || newParsed.hasRunningEntry;
    
    // Combine descriptions (unique only)
    const allDescriptions = [...existingParsed.descriptions, ...newParsed.descriptions];
    const uniqueDescriptions = Array.from(new Set(allDescriptions));
    
    // Format the merged entry
    return this.formatEntry(
      existingParsed.date,
      hours,
      uniqueDescriptions,
      hasRunningEntry
    );
  }
}

File: python/toggl_github_sync/formatter.py
"""Worklog formatting module."""

import calendar
import logging
import re
from datetime import datetime
from typing import List, Optional

import pytz

from toggl_github_sync.config import Config

logger = logging.getLogger(__name__)

class WorklogFormatter:
    """Formats worklog entries according to the required format."""

    def __init__(self, config: Config):
        """Initialize the worklog formatter.
        
        Args:
            config: Application configuration
        """
        self.config = config
        self.timezone = pytz.timezone(config.timezone)
    
    def format_entry(
        self, 
        date: datetime, 
        hours: float, 
        descriptions: List[str], 
        has_running_entry: bool
    ) -> str:
        """Format a worklog entry.
        
        Args:
            date: Date of the entry
            hours: Total hours worked
            descriptions: List of task descriptions
            has_running_entry: Whether there's a running entry
            
        Returns:
            Formatted worklog entry
        """
        # Format date as YYYY-M-D
        date_str = date.strftime("%Y-%-m-%-d")
        
        # Get day name (Mon, Tue, etc.)
        day_name = calendar.day_name[date.weekday()][:3]
        
        # Format hours as X.Yh or X.Yh+ if there's a running entry
        hours_str = f"{hours}h"
        if has_running_entry:
            hours_str += "+"
        
        # Join descriptions with periods
        description_text = ". ".join(descriptions)
        if description_text and not description_text.endswith("."):
            description_text += "."
        
        # Format the entry
        return f"{date_str} {day_name} ({hours_str}): {description_text}"
    
    def parse_entry(self, entry: str) -> Optional[dict]:
        """Parse a worklog entry.
        
        Args:
            entry: Worklog entry string
            
        Returns:
            Dictionary with parsed components or None if parsing failed
        """
        # Pattern to match worklog entries
        pattern = r"(\d{4}-\d{1,2}-\d{1,2}) ([A-Za-z]{3}) \((\d+\.?\d*)h(\+?)\): (.*)"
        match = re.match(pattern, entry)
        
        if not match:
            logger.warning(f"Failed to parse worklog entry: {entry}")
            return None
        
        date_str, day_name, hours_str, running_marker, description = match.groups()
        
        try:
            # Parse date
            year, month, day = map(int, date_str.split("-"))
            date = datetime(year, month, day, tzinfo=self.timezone)
            
            # Parse hours
            hours = float(hours_str)
            
            # Parse running status
            has_running_entry = running_marker == "+"
            
            # Parse descriptions
            descriptions = [d.strip() for d in description.split(".") if d.strip()]
            
            return {
                "date": date,
                "day_name": day_name,
                "hours": hours,
                "has_running_entry": has_running_entry,
                "descriptions": descriptions,
                "description_text": description,
            }
        except (ValueError, IndexError) as e:
            logger.warning(f"Error parsing worklog entry: {e}")
            return None
    
    def merge_entries(self, existing_entry: str, new_entry: str) -> str:
        """Merge an existing entry with a new entry.
        
        Args:
            existing_entry: Existing worklog entry
            new_entry: New worklog entry
            
        Returns:
            Merged worklog entry
        """
        existing_parsed = self.parse_entry(existing_entry)
        new_parsed = self.parse_entry(new_entry)
        
        if not existing_parsed or not new_parsed:
            logger.warning("Failed to parse entries for merging, using new entry")
            return new_entry
        
        # Use the higher hours value
        hours = max(existing_parsed["hours"], new_parsed["hours"])
        
        # Combine running status (if either is running, result is running)
        has_running_entry = existing_parsed["has_running_entry"] or new_parsed["has_running_entry"]
        
        # Combine descriptions (unique only)
        all_descriptions = existing_parsed["descriptions"] + new_parsed["descriptions"]
        unique_descriptions = []
        for desc in all_descriptions:
            if desc not in unique_descriptions:
                unique_descriptions.append(desc)
        
        # Format the merged entry
        return self.format_entry(
            existing_parsed["date"],
            hours,
            unique_descriptions,
            has_running_entry
        )
